compiler/src/lnntoamm/Types.ts

```diff
@@ -85,54 +103,85 @@ export class Builtin extends Type {
     return this;
   }
 
-  compatibleWithConstraint(ty: Type): boolean {
+  compatibleWithConstraint(ty: Type, scope: Scope): boolean {
     if (ty instanceof Builtin) {
       return this.eq(ty);
-    } if (ty instanceof OneOf || ty instanceof Generated) {
-      return ty.compatibleWithConstraint(this);
+    } else if (ty instanceof OneOf || ty instanceof Generated) {
+      return ty.compatibleWithConstraint(this, scope);
+    } else if (ty instanceof HasOperator) {
+      return Has.operator(ty, scope, this).length !== 0;
+    } else if (ty instanceof HasMethod) {
+      return Has.method(ty, scope, this).length !== 0;
+    } else if (ty instanceof HasField) {
+      return false;
+    } else if (ty instanceof Interface) {
+      return ty.fields.length === 0 &&
+            ty.methods.every(m => this.compatibleWithConstraint(m, scope)) &&
+            ty.operators.every(o => this.compatibleWithConstraint(o, scope));
     } else {
       TODO('other types constraining to builtin types');
     }
   }
 
-  constrain(ty: Type) {
-    if (ty instanceof OneOf || ty instanceof Generated) {
-      ty.constrain(this);
-    } else if (!this.compatibleWithConstraint(ty)) {
+  constrain(ty: Type, scope: Scope) {
+    if (ty instanceof OneOf || ty instanceof Generated || ty instanceof Interface) {
+      ty.constrain(this, scope);
+    } else if (ty instanceof HasOperator) {
+      if (Has.operator(ty, scope, this).length === 0) {
+        throw new Error(`type ${this.name} does not have operator ${ty.name}`);
+      }
```

Comment:
Nit. You can combine the nested conditional up without impacting the logic.

```suggestion
    } else if (ty instanceof HasOperator && Has.operator(ty, scope, this).length === 0) {
      throw new Error(`type ${this.name} does not have operator ${ty.name}`);
```

Reply:
I actually can't collapse this once since `HasOperator extends HasMethod` and will get checked by `Has.method` on the next branch, but i can combine the `Has.method` conditional with `instanceof HasMethod`

Reply:
Ahh. Why does `HasOperator` extend `HasMethod`, btw? I wouldn't have expected there to be that much code reuse between the two?

Reply:
for 2 reasons: first, that operators are a kind of methods and so it makes sense to me, but also because that way the operators inherit the `params` and `ret` fields from `HasMethod` and can inherit behavior around the 2. I'm open to changing it, but this is what made sense to me

Reply:
No, it's minor; I just probably wouldn't have done it that way, but I don't see any harm in it?